// @squidcloud/local-backend/dist/internal-common/src/types/communication.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.omitSquidDevId = exports.verifyWithSquidDevId = exports.validateEnvironment = exports.appIdWithEnvironmentIdAndDevId = exports.appIdWithEnvironmentId = exports.parseAppId = void 0;
const assertic_1 = require("assertic");
function parseAppId(appId) {
    (0, assertic_1.assertString)(appId, 'Invalid application ID');
    const [appIdWithoutEnv, environmentId, squidDeveloperId, other] = appId.split('-');
    (0, assertic_1.assertTruthy)(!other, `Invalid application ID: ${appId}`);
    return {
        appId: appIdWithoutEnv,
        environmentId: (environmentId !== null && environmentId !== void 0 ? environmentId : 'prod'),
        squidDeveloperId,
    };
}
exports.parseAppId = parseAppId;
function appIdWithEnvironmentId(appId, environmentId) {
    const parsedAppId = parseAppId(appId);
    return `${parsedAppId.appId}${!environmentId || environmentId === 'prod' ? '' : `-${environmentId}`}`;
}
exports.appIdWithEnvironmentId = appIdWithEnvironmentId;
function appIdWithEnvironmentIdAndDevId(appId, environmentId, developerId) {
    return `${appIdWithEnvironmentId(appId, environmentId)}${developerId ? `-${developerId}` : ''}`;
}
exports.appIdWithEnvironmentIdAndDevId = appIdWithEnvironmentIdAndDevId;
function validateEnvironment(appId) {
    const parsedAppId = parseAppId(appId);
    (0, assertic_1.assertTruthy)(parsedAppId.environmentId !== 'prod' || !parsedAppId.squidDeveloperId, 'Cannot use developer ID in production');
    return parsedAppId;
}
exports.validateEnvironment = validateEnvironment;
function verifyWithSquidDevId(appId) {
    const parsedAppId = parseAppId(appId);
    (0, assertic_1.assertTruthy)(parsedAppId.squidDeveloperId, 'Must provide squid developer ID');
}
exports.verifyWithSquidDevId = verifyWithSquidDevId;
function omitSquidDevId(appId) {
    const parsedAppId = parseAppId(appId);
    return appIdWithEnvironmentId(parsedAppId.appId, parsedAppId.environmentId);
}
exports.omitSquidDevId = omitSquidDevId;
