// @squidcloud/local-backend/dist/internal-common/src/public-types/ai-chatbot.public-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AiChatbotChatContext = exports.AI_IMAGE_MODEL_NAMES = exports.STABLE_DIFFUSION_MODEL_NAMES = exports.OPENAI_IMAGE_MODEL_NAMES = exports.AI_EMBEDDINGS_MODEL_NAMES = exports.OPENAI_EMBEDDINGS_MODEL_NAMES = exports.AI_CHAT_MODEL_NAMES = exports.ANTHROPIC_CHAT_MODEL_NAMES = exports.GEMINI_CHAT_MODEL_NAMES = exports.OPENAI_CHAT_MODEL_NAMES = void 0;
exports.OPENAI_CHAT_MODEL_NAMES = [
    'gpt-3.5-turbo',
    'gpt-3.5-turbo-1106',
    'gpt-4',
    'gpt-4-turbo-preview',
    'gpt-4o',
    'gpt-4-turbo',
];
exports.GEMINI_CHAT_MODEL_NAMES = ['gemini-pro'];
exports.ANTHROPIC_CHAT_MODEL_NAMES = [
    'claude-3-opus-20240229',
    'claude-3-sonnet-20240229',
    'claude-3-haiku-20240307',
    'claude-3-5-sonnet-20240620',
];
exports.AI_CHAT_MODEL_NAMES = [
    ...exports.OPENAI_CHAT_MODEL_NAMES,
    ...exports.ANTHROPIC_CHAT_MODEL_NAMES,
    ...exports.GEMINI_CHAT_MODEL_NAMES,
];
exports.OPENAI_EMBEDDINGS_MODEL_NAMES = [
    'text-embedding-3-small',
    'text-embedding-3-large',
    'text-embedding-ada-002',
];
exports.AI_EMBEDDINGS_MODEL_NAMES = [...exports.OPENAI_EMBEDDINGS_MODEL_NAMES];
exports.OPENAI_IMAGE_MODEL_NAMES = ['dall-e-3'];
exports.STABLE_DIFFUSION_MODEL_NAMES = ['stable-diffusion-core'];
exports.AI_IMAGE_MODEL_NAMES = [...exports.OPENAI_IMAGE_MODEL_NAMES, ...exports.STABLE_DIFFUSION_MODEL_NAMES];
class AiChatbotChatContext {
    constructor(prototype) {
        this.profileId = prototype.profileId;
        this.prompt = prototype.prompt;
    }
}
exports.AiChatbotChatContext = AiChatbotChatContext;
