// @squidcloud/local-backend/dist/internal-common/src/public-types/integration.public-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegrationSchemaType = exports.IntegrationType = void 0;
var IntegrationType;
(function (IntegrationType) {
    IntegrationType["built_in_db"] = "built_in_db";
    IntegrationType["mongo"] = "mongo";
    IntegrationType["mysql"] = "mysql";
    IntegrationType["clickhouse"] = "clickhouse";
    IntegrationType["mssql"] = "mssql";
    IntegrationType["postgres"] = "postgres";
    IntegrationType["cockroach"] = "cockroach";
    IntegrationType["api"] = "api";
    IntegrationType["graphql"] = "graphql";
    IntegrationType["snowflake"] = "snowflake";
    IntegrationType["datadog"] = "datadog";
    IntegrationType["newrelic"] = "newrelic";
    IntegrationType["auth0"] = "auth0";
    IntegrationType["jwt_rsa"] = "jwt_rsa";
    IntegrationType["jwt_hmac"] = "jwt_hmac";
    IntegrationType["ai_chatbot"] = "ai_chatbot";
    IntegrationType["cognito"] = "cognito";
    IntegrationType["okta"] = "okta";
    IntegrationType["descope"] = "descope";
    IntegrationType["firebase_auth"] = "firebase_auth";
    IntegrationType["kafka"] = "kafka";
    IntegrationType["confluent"] = "confluent";
    IntegrationType["built_in_queue"] = "built_in_queue";
    IntegrationType["s3"] = "s3";
    IntegrationType["gcs"] = "gcs";
    IntegrationType["built_in_s3"] = "built_in_s3";
    IntegrationType["built_in_gcs"] = "built_in_gcs";
    IntegrationType["algolia"] = "algolia";
    IntegrationType["elastic_observability"] = "elastic_observability";
    IntegrationType["elastic_search"] = "elastic_search";
    IntegrationType["elastic_enterprise_search"] = "elastic_enterprise_search";
    IntegrationType["sentry"] = "sentry";
    IntegrationType["sap_hana"] = "sap_hana";
    IntegrationType["salesforce_crm"] = "salesforce_crm";
    IntegrationType["documentdb"] = "documentdb";
    IntegrationType["dynamodb"] = "dynamodb";
    IntegrationType["cassandra"] = "cassandra";
    IntegrationType["alloydb"] = "alloydb";
    IntegrationType["spanner"] = "spanner";
    IntegrationType["db2"] = "db2";
    IntegrationType["mariadb"] = "mariadb";
    IntegrationType["oracledb"] = "oracledb";
    IntegrationType["redis"] = "redis";
    IntegrationType["xata"] = "xata";
    IntegrationType["azure_sql"] = "azure_sql";
    IntegrationType["azure_postgresql"] = "azure_postgresql";
    IntegrationType["azure_cosmosdb"] = "azure_cosmosdb";
    IntegrationType["firestore"] = "firestore";
    IntegrationType["bigquery"] = "bigquery";
    IntegrationType["cloudsql"] = "cloudsql";
})(IntegrationType || (exports.IntegrationType = IntegrationType = {}));
var IntegrationSchemaType;
(function (IntegrationSchemaType) {
    IntegrationSchemaType["data"] = "data";
    IntegrationSchemaType["api"] = "api";
    IntegrationSchemaType["graphql"] = "graphql";
})(IntegrationSchemaType || (exports.IntegrationSchemaType = IntegrationSchemaType = {}));
