// @squidcloud/local-backend/dist/local-backend/src/local-backend.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var LocalBackendModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalBackendModule = exports.BodyInterceptor = void 0;
const common_1 = require("@nestjs/common");
const local_backend_service_1 = require("./local-backend.service");
const local_backend_controller_1 = require("./local-backend.controller");
const local_backend_socket_service_1 = require("./local-backend-socket.service");
const config_1 = require("@nestjs/config");
const serialization_1 = require("../../internal-common/src/utils/serialization");
let BodyInterceptor = class BodyInterceptor {
    async intercept(ctx, next) {
        if (ctx.getType() !== 'http')
            return next.handle();
        const httpRequest = ctx.switchToHttp().getRequest();
        httpRequest.body = httpRequest.rawBody ? (0, serialization_1.deserializeObj)(httpRequest.rawBody.toString()) : {};
        return next.handle();
    }
};
exports.BodyInterceptor = BodyInterceptor;
exports.BodyInterceptor = BodyInterceptor = __decorate([
    (0, common_1.Injectable)()
], BodyInterceptor);
let LocalBackendModule = LocalBackendModule_1 = class LocalBackendModule {
    static forRoot(backendService) {
        const providers = [
            {
                provide: 'BACKEND_SERVICE',
                useValue: backendService,
            },
        ];
        return {
            module: LocalBackendModule_1,
            global: true,
            providers: providers,
            exports: providers,
        };
    }
};
exports.LocalBackendModule = LocalBackendModule;
exports.LocalBackendModule = LocalBackendModule = LocalBackendModule_1 = __decorate([
    (0, common_1.Module)({
        imports: [config_1.ConfigModule.forRoot()],
        controllers: [local_backend_controller_1.LocalBackendController],
        providers: [local_backend_service_1.LocalBackendService, local_backend_socket_service_1.LocalBackendSocketService, BodyInterceptor],
    })
], LocalBackendModule);
