// assertic/dist/src/Validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateRecord = exports.validateArray = exports.validateObject = exports.tryCatch = void 0;
const Assertion_1 = require("./Assertion");
/**
 * Runs a function guarded by try/catch and returns an error message if the function throws an error.
 * Returns 'undefined' if the function run with no errors.
 */
function tryCatch(fn) {
    try {
        fn();
    }
    catch (error) {
        return (error instanceof Error && error.message) || `${error}`;
    }
}
exports.tryCatch = tryCatch;
function validateObject(value, assertion, errorContextProvider = undefined, constraints = {}) {
    return tryCatch(() => (0, Assertion_1.assertObject)(value, assertion, errorContextProvider, constraints));
}
exports.validateObject = validateObject;
function validateArray(value, elementAssertion, constraints = {}, errorContextProvider = undefined) {
    return tryCatch(() => (0, Assertion_1.assertArray)(value, elementAssertion, constraints, errorContextProvider));
}
exports.validateArray = validateArray;
function validateRecord(value, valueAssertion, constraints = {}, errorContextProvider = undefined) {
    return tryCatch(() => (0, Assertion_1.assertRecord)(value, valueAssertion, constraints, errorContextProvider));
}
exports.validateRecord = validateRecord;
