// @netlify/content-engine/dist/reporter/constants.js
"use strict";

exports.__esModule = true;
exports.LogLevels = exports.ActivityTypes = exports.ActivityStatuses = exports.ActivityLogLevels = exports.Actions = void 0;
let Actions = exports.Actions = /*#__PURE__*/function (Actions) {
  Actions["LogAction"] = "LOG_ACTION";
  Actions["SetStatus"] = "SET_STATUS";
  Actions["Log"] = "LOG";
  Actions["SetLogs"] = "SET_LOGS";
  Actions["StartActivity"] = "ACTIVITY_START";
  Actions["EndActivity"] = "ACTIVITY_END";
  Actions["UpdateActivity"] = "ACTIVITY_UPDATE";
  Actions["PendingActivity"] = "ACTIVITY_PENDING";
  Actions["CancelActivity"] = "ACTIVITY_CANCEL";
  Actions["ActivityErrored"] = "ACTIVITY_ERRORED";
  return Actions;
}({});
let LogLevels = exports.LogLevels = /*#__PURE__*/function (LogLevels) {
  LogLevels["Debug"] = "DEBUG";
  LogLevels["Success"] = "SUCCESS";
  LogLevels["Info"] = "INFO";
  LogLevels["Warning"] = "WARNING";
  LogLevels["Log"] = "LOG";
  LogLevels["Error"] = "ERROR";
  return LogLevels;
}({});
let ActivityLogLevels = exports.ActivityLogLevels = /*#__PURE__*/function (ActivityLogLevels) {
  ActivityLogLevels["Success"] = "ACTIVITY_SUCCESS";
  ActivityLogLevels["Failed"] = "ACTIVITY_FAILED";
  ActivityLogLevels["Interrupted"] = "ACTIVITY_INTERRUPTED";
  return ActivityLogLevels;
}({});
let ActivityStatuses = exports.ActivityStatuses = /*#__PURE__*/function (ActivityStatuses) {
  ActivityStatuses["InProgress"] = "IN_PROGRESS";
  ActivityStatuses["NotStarted"] = "NOT_STARTED";
  ActivityStatuses["Interrupted"] = "INTERRUPTED";
  ActivityStatuses["Failed"] = "FAILED";
  ActivityStatuses["Success"] = "SUCCESS";
  ActivityStatuses["Cancelled"] = "CANCELLED";
  return ActivityStatuses;
}({});
let ActivityTypes = exports.ActivityTypes = /*#__PURE__*/function (ActivityTypes) {
  ActivityTypes["Spinner"] = "spinner";
  ActivityTypes["Hidden"] = "hidden";
  ActivityTypes["Progress"] = "progress";
  ActivityTypes["Pending"] = "pending";
  return ActivityTypes;
}({});
